### PR TITLE
[ENG-3474] feat: Make UI for STX RBF in the extension

### PR DESCRIPTION
**Pull request by dhriaznov**
# 🔘 PR Type
What kind of change does this PR introduce?

- [x] Enhancement

# 📜 Background
This feature is needed to help our Stacks users increase the chance of getting their pending transactions included in the next Stacks anchor block, by increasing the transaction fee.

Issue Link: #[[ENG-3474](https://linear.app/xverseapp/issue/ENG-3474/make-ui-for-stx-rbf-in-the-extension)]
Context Link (if applicable):

# 🔄 Changes
- Added `useRbfTransactionData` hook
- Updated the Speed Up Tx page to add the STX functionality
- Added the Speed Up button to `src/app/components/transactions/stxTransferTransaction.tsx` component
- Removed the unused code in `src/app/components/transactions/stxTransaction.tsx` component
- Removed the unused `src/assets/img/transactions/increaseFee.svg` icon

Impact:
- This PR impacts both BTC and STX RBF features

# 🖼 Screenshot / 📹 Video
The demo videos are in the linear ticket.

Monosnap Monosnap 2023-12-28 17-00-19
Monosnap Monosnap 2023-12-28 17-00-27
Monosnap Monosnap 2023-12-28 17-00-37
Monosnap Monosnap 2023-12-28 17-00-59

# ✅ Review checklist
Please ensure the following are true before merging:

- [x] Code Style is consistent with the project guidelines.
- [ ] Code is readable and well-commented.
- [ ] No unnecessary or debugging code has been added.
- [x] Security considerations have been taken into account.
- [ ] The change has been manually tested and works as expected.
- [x] Breaking changes and their impacts have been considered and documented.
- [x] Code does not introduce new technical debt or issues.

### CODE REVIEW

File: src/app/hooks/useRbfTransactionData.ts
Comment (teebszet) [inline src/app/hooks/useRbfTransactionData.ts:132]: nit: could use .lte() rather than double negative?

File: src/app/hooks/useRbfTransactionData.ts
Comment (teebszet) [inline src/app/hooks/useRbfTransactionData.ts:119]: is there an edge case here, where the minimumFee can be higher than the medium/high Fees?

File: src/app/screens/speedUpTransaction/index.tsx
Comment (teebszet) [inline src/app/screens/speedUpTransaction/index.tsx:1]: @dhriaznov IMO we should stop writing new code with this pattern of shared screens/large components between STX/BTC 
`if(isBtc) { // do btc } else { // do stacks }`

it makes the screen file bloated, and more difficult to maintain/change in future (e.g. see ConfirmBtcTransactionComponent.tsx, or how sendForm.tsx was eventually extended to support many currencyTypes). 

If we have the time, I would much prefer to duplicate the screen, and reuse the lower level components, or reuse the layout containers

File: src/app/hooks/useRbfTransactionData.ts
Comment (dhriaznov) [reply src/app/hooks/useRbfTransactionData.ts:119]: yeah, I think that there could be this edge case

File: src/app/screens/speedUpTransaction/index.tsx
Comment (dhriaznov) [reply src/app/screens/speedUpTransaction/index.tsx:1]: I agree, I created separate components now

File: src/app/hooks/useRbfTransactionData.ts
Comment (teebszet) [reply src/app/hooks/useRbfTransactionData.ts:120]: > yeah, I think that there could be this edge case. should we disable those options then or calculate them to be higher than the current fee?

the unit tests and logic should follow the user story set out here:

https://www.notion.so/xverseapp/Speed-up-Stacks-transaction-Improve-fee-estimation-4455d27f8bdc4ebfbd94a16d7f18fe12?pvs=4#6e0e607014764f2ab9d6f8b76d677dfe

Comment (teebszet) [reply src/app/hooks/useRbfTransactionData.ts:120]: for [SPEEDSTX-EXT-2]

Comment (teebszet) [reply src/app/hooks/useRbfTransactionData.ts:120]: to clarify: I don't mind when you move logic into core in this case, since you're going to be working on both clients anyway.

but before we can merge, the feature should satisfy the user stories set out in notion.

since we have really clear requirements/specs now, the most efficient way would be for  the PR author to quickly double check the PR matches the notion (in case the requirements changed or no longer make sense, etc.)

thanks!

Comment (dhriaznov) [reply src/app/hooks/useRbfTransactionData.ts]: totally makes sense @teebszet, thanks for noticing that! I updated the logic to follower the user story Case 2 and Case 3
1 Monosnap Monosnap 2024-01-08 18-24-35
2 Monosnap Monosnap 2024-01-08 18-25-06

File: src/app/hooks/useRbfTransactionData.ts
Comment (teebszet) [inline src/app/hooks/useRbfTransactionData.ts:232]: it kinda feels like this should be two separate hooks. 

you end up loading more code than you need to, like btcClient (when doing stx rbf) and selectedNetwork, network, stxAvailableBalance (when doing btc rbf).

did you consider splitting the screens/speedUpTransaction/index.tsx into btc and stx versions? I saw you split it up on the component level instead

File: src/app/hooks/useRbfTransactionData.ts
Comment (teebszet) [inline src/app/hooks/useRbfTransactionData.ts:157]: what was the requirements for this part?

should you not do all calculations on microstacks, and only convert to stx if you need to before/in the return statement?

File: src/app/hooks/useRbfTransactionData.ts
Comment (teebszet) [inline src/app/hooks/useRbfTransactionData.ts:198]: 👍

File: src/app/hooks/useRbfTransactionData.ts
Comment (teebszet) [inline src/app/hooks/useRbfTransactionData.ts:80]: does this need to be async?

File: src/app/hooks/useRbfTransactionData.ts
Comment (dhriaznov) [reply src/app/hooks/useRbfTransactionData.ts:157]: the requirement for the minimum fee is here https://secretkeylabs.slack.com/archives/C063ZVDLCLB/p1703778799805989?thread_ts=1703712686.356469&cid=C063ZVDLCLB

File: src/app/hooks/useRbfTransactionData.ts
Comment (dhriaznov) [reply src/app/hooks/useRbfTransactionData.ts:80]: nope, it doesn't, good gatch 🙂

File: src/app/hooks/useRbfTransactionData.ts
Comment (dhriaznov) [reply src/app/hooks/useRbfTransactionData.ts:232]: that's a valid point.

regarding separate hooks - my suggestion here is to see how it'll look like when we move the helper functions and some of the STX logic to the core repo and then decide.

and regarding separate screens - not sure if it's necessary since they share quite a lot of logic and UI currently. if we make the separate hooks then we probably should separate the screens as well

Comment (teebszet) [reply src/app/hooks/useRbfTransactionData.ts:232]: > if we make the separate hooks then we probably should separate the screens as well

yeah that would make sense

> regarding separate hooks - my suggestion here is to see how it'll look like when we move the helper functions and some of the STX logic to the core repo and then decide.

OK

File: src/app/hooks/useRbfTransactionData.ts
Comment (teebszet) [inline src/app/hooks/useRbfTransactionData.ts:153]:
```suggestion
      const currentMicrostacksFee = transaction.fee;
```

yeah the minimum fee rounding is a lot cleaner, thanks!

nit: you still have this extra arithmetic

File: src/app/hooks/useRbfTransactionData.ts
Comment (teebszet) [inline src/app/hooks/useRbfTransactionData.ts:143]: const fee is not used until L203 L204?

```suggestion
```

File: src/app/hooks/useRbfTransactionData.ts
Comment (teebszet) [inline src/app/hooks/useRbfTransactionData.ts:204]:
```suggestion
          currentFee: microStxToStx(currentMicrostacksFee).toNumber(),
          currentFeeRate: microStxToStx(currentMicrostacksFee).toNumber(),
```

File: src/app/hooks/useRbfTransactionData.ts
Comment (teebszet) [reply src/app/hooks/useRbfTransactionData.ts]: thanks! much cleaner

**Review by teebszet — APPROVED** [review]

File: src/app/hooks/useRbfTransactionData.ts
Comment (teebszet) [inline src/app/hooks/useRbfTransactionData.ts:12]: nit: xverse-core already exports stxToMicrostacks and microstacksToStx
```suggestion
  stxToMicrostacks,
  microstacksToStx
  StxTransactionData,
} from '@secretkeylabs/xverse-core';
import { deserializeTransaction, estimateTransaction } from '@stacks/transactions';
import { isLedgerAccount } from '@utils/helper';
```

**Comment by teebszet** [comment]
@dhriaznov thanks for make the code improvements. let's merge this after the base branch is merged

**Comment by dhriaznov** [comment]
Thanks for the review @teebszet! I resolved the git conflicts and applied the above code suggestions

**Review by teebszet — DISMISSED** [review]

**Comment by DuskaT021** [comment]
@dhriaznov needs conflicts resolved

**Review by teebszet — APPROVED** [review]

**Pull request merged**
